### PR TITLE
Add snapshot compatibility checks for snapshot create/restore

### PR DESCRIPTION
**Pull request by georgepisaltu**
# Reason for This PR

VMs restored from a snapshot might have the TSC frequency misaligned with the original, which could lead to problems further down the line. We took this as an opportunity to add snapshot compatibility checks, which start with adding the original TSC in the state file and scaling it for resumed VMs.

## Description of Changes

Added compatibility checks for snapshot create/restore, starting with TSC scaling.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.

### CODE REVIEW

File: src/api_server/swagger/firecracker.yaml [hunk 1]
```diff
@@ -1018,6 +1022,10 @@ definitions:
         type: boolean
         description:
           When set to true, the vm is also resumed if the snapshot load is successful.
+      skip_compatibility_check:
+        type: boolean
+        description:
+          Optional field to skip the snapshot compatibility checks, which are on by default.
```
Comment (acatangiu) [inline src/api_server/swagger/firecracker.yaml:1028]: Why do we need this knob? It adds a lot of complexity on the user/usage side as well Firecracker code complexity for handling all new cases introduced by it.

If the only production-worthy reason for it is to allow loading old snapshots, I suggest we look at some other approaches for migrating/upgrading those snaps, instead of adding and maintaining an extra knob in perpetuity..

Comment (acatangiu) [reply src/api_server/swagger/firecracker.yaml:1028]: We could offer a dedicated snapshot compatibility tool for example, or simply warn when coming from TSC-unaware snaps but do best-effort and load them.

**Comment by dianpopa** [comment]
Should we close [this one](https://github.com/firecracker-microvm/firecracker/pull/2522/files)?

File: src/api_server/swagger/firecracker.yaml (hunk 1 above)
Comment (berciuliviu) [reply src/api_server/swagger/firecracker.yaml:1028]: I agree with Adrian, we should take a simpler approach to this issue, without adding a new parameter into the account. I also believe that validating v0.25 snapshots for tsc_freq should be done by default, while loading an older-version snapshot should just warn of tsc_freq not available. If my understanding is correct, vcpu_state.tsc_freq is automatically set to None if an older version snapshot is loaded.

File: src/cpuid/src/common.rs [hunk 2]
```diff
@@ -91,6 +92,39 @@ pub fn get_vendor_id_from_cpuid(cpuid: &CpuId) -> Result<[u8; 12], Error> {
     Err(Error::NotSupported)
 }
 
+/// Validates that a CPUID and another leaf 0x1 describe the same
+/// CPU in terms of manufacturer and model.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn is_same_model(cpuid: &CpuId, other_leaf0x1_eax: u32) -> bool {
```
Comment (serban300) [inline src/cpuid/src/common.rs:98]: I wouldn't supply the other `0x1` leaf as a parameter. I would read it inside this method.

File: src/cpuid/src/common.rs [hunk 3]
```diff
@@ -91,6 +92,39 @@ pub fn get_vendor_id_from_cpuid(cpuid: &CpuId) -> Result<[u8; 12], Error> {
...
+    // Search for the entry for leaf0x1.
+    let leaf0x1 = cpuid
```
Comment (serban300) [inline src/cpuid/src/common.rs:100]: The `0x1` leaf is also referred to as the feature information leaf. So I think we can name this something along the lines of `feature_info_leaf`.

File: src/cpuid/src/common.rs [hunk 4]
```diff
@@ -91,6 +92,39 @@ pub fn get_vendor_id_from_cpuid(cpuid: &CpuId) -> Result<[u8; 12], Error> {
...
+
+    // Validate that all of these properties are the same.
+    for elem in &[
+        leaf_0x1::eax::EXTENDED_FAMILY_ID_BITRANGE,
+        leaf_0x1::eax::EXTENDED_PROCESSOR_MODEL_BITRANGE,
+        leaf_0x1::eax::PROCESSOR_FAMILY_BITRANGE,
+        leaf_0x1::eax::PROCESSOR_MODEL_BITRANGE,
+        leaf_0x1::eax::STEPPING_BITRANGE,
+    ] {
+        if leaf0x1_eax.read_bits_in_range(elem) != other_leaf0x1_eax.read_bits_in_range(elem) {
```
Comment (serban300) [inline src/cpuid/src/common.rs:120]: Are this values guaranteed to be unique even across different vendors ? Shouldn't we also compare the vendor just to make sure ? The vendor can be extracted from the brand string and we might already have a helper method for that.

File: src/cpuid/src/common.rs [hunk 5]
```diff
@@ -91,6 +92,39 @@ pub fn get_vendor_id_from_cpuid(cpuid: &CpuId) -> Result<[u8; 12], Error> {
...
+        .as_slice()
+        .iter()
+        .find(|entry| entry.function == leaf_0x1::LEAF_NUM);
+    // Fail fast if we can't retrieve the relevant information from CPUID.
+    if leaf0x1.is_none() {
+        return false;
+    }
+
+    // The relevant information is in EAX.
+    let leaf0x1_eax = leaf0x1.unwrap().eax;
```
Comment (serban300) [inline src/cpuid/src/common.rs:110]: Does something like `let leaf0x1_eax = leaf0x1?.eax;` work ?

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 6]
```diff
@@ -291,9 +307,55 @@ impl KvmVcpu {
...
             xsave,
+            tsc_khz,
         })
     }
 
+    // Scale the TSC frequency of this vCPU to the one provided as a parameter.
+    pub fn set_tsc_khz(&self, state_tsc: Option<u32>, cpuid: &CpuId) -> Result<()> {
+        // Check if CPU models are the same.
+        if let Ok(curr_cpuid) = cpuid::common::get_cpuid(0x1, 0) {
+            if is_same_model(cpuid, curr_cpuid.eax) {
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:318]: This allow snapshots to be loaded on the same CPU, even if there is no constant TSC or TSC scaling support. Are we sure the checks we do here guarantee that TSC freq is the same ?

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 7]
```diff
@@ -291,9 +307,55 @@ impl KvmVcpu {
...
+        } else {
+            state_tsc.unwrap()
+        };
+
+        match curr_kvm_tsc {
+            Ok(curr_tsc_freq) => {
+                // Compare the current TSC freq to the one found
+                // in the state. If they are different, we need to
+                // scale the TSC to the freq found in the state.
+                if curr_tsc_freq != snapshot_tsc_freq {
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:340]: I think we should allow for some margin of error here, since even on the same CPU you might get different values during calibration.

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 8]
```diff
@@ -392,6 +457,30 @@ pub struct VcpuState {
...
+        None
+    }
+
+    fn ser_tsc(&mut self, target_version: u16) -> VersionizeResult<()> {
+        if target_version < 2 {
+            warn!("Saving to older snapshot version, TSC freq not included in snapshot.");
+        }
+
+        // Prevent v0.25 to create snapshots of uVMs running on non-invariant TSC host.
+        if self.tsc_khz.is_none() {
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:476]: We dont't allow this but we do load snaps from v0.24 (without TSC) given the CPUID model information matches. What happens if we drop this check ?

File: src/cpuid/src/common.rs (hunk 4 above)
Comment (georgepisaltu) [reply src/cpuid/src/common.rs:120]: My research led me to believe they are unique across every CPU model, which means different vendors would make them different as well.

However, the vendor string check doesn't hurt in this case so I added it just to be safe.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 6 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:318]: If we as a process never change the TSC frequency outside the scope of a snapshot save/restore, and the CPU model is the same, I expect the TSC frequency to be the same in the clone as in the original VM. If this is not true, then checking the CPU model is actually useless, since we still need to get/set the TSC frequency regardless.

On variant TSC systems, I don't think users even expect high precision from the TSC, but I might have a bad understanding of this whole story.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 7 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:340]: Any margin of error here can lead to a great divergence in TSC over a period of time.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 8 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:476]: Depends on whether you want to load the TSC-less snapshot on the same CPU model or not. I guess this ties in to the previous comment.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 7 above)
Comment (sandreim) [reply src/vmm/src/vstate/vcpu/x86_64.rs:339]: Some differences can be corrected by the guest NTP, more info in QEMU approach: https://patchew.org/QEMU/20200616165805.GA324612@fuller.cnet/

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 8 above)
Comment (sandreim) [reply src/vmm/src/vstate/vcpu/x86_64.rs:475]: I think we should have a consistent behavior both for load and save wrt older snapshot versions.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 7 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:339]: Updated it/

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 8 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:475]: Dropped the check, now the failure path is on snapshot restore. We still warn users about this with a log.

**Review by georgepisaltu — COMMENTED** [review]
@sandreim PTAL!

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 9]
```diff
@@ -291,9 +309,57 @@ impl KvmVcpu {
...
     }
 
+    // Scale the TSC frequency of this vCPU to the one provided as a parameter.
+    pub fn set_tsc_khz(&self, state_tsc: Option<u32>, cpuid: &CpuId) -> Result<()> {
+        // Check if CPU models are the same.
+        if is_same_model(cpuid) {
+            // If they are the same, we don't need to do anything else.
+            return Ok(());
+        }
+
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:323]: Should we fail early here if KVM caps for get/set TSC freq are not available ?

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 10]
```diff
@@ -250,6 +266,8 @@ impl KvmVcpu {
          * meaningful. For SET_MSRS it will then contain good data.
          */
 
+        let tsc_khz = self.get_tsc_khz()?;
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:269]: Do we still want to allow snapshots to be created if this fails ?

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 11]
```diff
@@ -392,6 +461,34 @@ pub struct VcpuState {
...
+    fn default_tsc_khz(_: u16) -> Option<u32> {
+        warn!("CPU TSC freq not found in snapshot");
+        None
+    }
+
+    fn ser_tsc(&mut self, target_version: u16) -> VersionizeResult<()> {
+        if target_version < 2 {
+            // v0.24 and older versions do not support TSC scaling.
+            warn!(
+                "Saving to older snapshot version, TSC freq not included \
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:478]: It would be helpful to also include the TSC freq in the message.

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 12]
```diff
@@ -392,6 +461,34 @@ pub struct VcpuState {
...
+                in snapshot."
+            );
+        } else if self.tsc_khz.is_none() {
+            // v0.25 and newer snapshots without TSC will only work on
+            // the same CPU model as the host on which they were taken.
+            warn!(
+                "TSC freq not available. Snapshot will only work on the \
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:485]: I would rephrase: `Snapshot cannot be loaded on a different CPU model.`

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 13]
```diff
@@ -291,9 +309,57 @@ impl KvmVcpu {
...
+    // Scale the TSC frequency of this vCPU to the one provided as a parameter.
+    pub fn set_tsc_khz(&self, state_tsc: Option<u32>, cpuid: &CpuId) -> Result<()> {
+        // Check if CPU models are the same.
+        if is_same_model(cpuid) {
+            // If they are the same, we don't need to do anything else.
+            return Ok(());
+        }
+
+        // Try to get a reading of the current TSC freq.
+        let curr_kvm_tsc = self.fd.get_tsc_khz();
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:325]: Can we refactor so the syscall and validation get called only once ? I expect the TSC freq to be the same across all vCPUs, so the load hotpath can be simplified.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 10 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:269]: Yes, as we discussed earlier. Fixed it.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 9 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:323]: It is a rather complex state where the `KVM_GET_TSC_KHZ` cap is always needed at this point, but the call is made just below and a failure does not tear the whole restore process down. Furthermore, we will only know if we need the `KVM_SET_TSC_KHZ` cap later, and we check it when we need it. So I don't really see the point of checking the capabilities here.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 13 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:325]: I found it very difficult to do this design-wise. There is no good place to do this, as this is in the snap restore path and the "owners" of the VCPUs are, in order, the `builder.rs` function `build_microvm_from_snapshot` and then the VCPU threads themselves, which communicate with the VMM via events. It would be really awkward to do it by passing parameters through the event enum. The other option is to do it with the builder, but there is no good place to do it there either.

I tested this POC and there seems to be no performance degradation in snapshot restore times. If we do the validations once before the VCPUs are moved to their threads or if we do it on every VCPU thread, the total amount of time to restore completion should be roughly the same. Let me know if I missed anything and we need to revisit this.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 11 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:478]: Done.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 12 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:485]: Done.

**Review by georgepisaltu — COMMENTED** [review]
@sandreim @serban300 I updated the PR, also added unit tests.

The next step is to add integration tests using snapshots generated with the final form of the binary after merging this.

PTAL!

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 13 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs]: I thought about it more and I attempted the refactor in the current iteration of the PR. Let me know if it makes sense. After all, doing the checks only once is the right way to go forward in the long run.

File: src/vmm/src/builder.rs [hunk 14]
```diff
@@ -427,6 +427,27 @@ pub fn build_microvm_from_snapshot(
...
+    {
+        // Scale the TSC frequency for all VCPUs.
+        for vcpu in &vcpus {
+            vcpu.kvm_vcpu
+                .set_tsc_khz(microvm_state.vcpu_states[0].tsc_khz)
+                .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
+                .map_err(RestoreMicrovmState)?;
+        }
+    }
+
```
Comment (serban300) [inline src/vmm/src/builder.rs:450]: This approach leads to some developments that I personally find  both unsafe and hard to follow. For example:

```
    pub fn set_tsc_khz(&self, state_tsc: Option<u32>) -> Result<()> {
        // Extract the TSC freq from the state.
        let snapshot_tsc_freq = match state_tsc {
            Some(freq) => freq,
            None => {
                // The snapshot TSC was already validated.
                unreachable!();
            }
        };
        ...
    }
```
Here relying on the fact that the TSC was already validated is very risky since it's not in the control of the function and not even in the control of the class. `state_tsc` is a parameter that we receive and as far as we know here, it's completely out of our control.

I think this is all generated by the attempt to squeeze the `is_same_model()` check inside the `is_tsc_scaling_required()`, which definitely makes sense initially, but it is not worth the tradeoffs. As an incremental improvement here, just to make the code safer, I would at least separate `is_same_mode()` from `is_tsc_scaling_required()`, changing this code to something like:

```
        if !cpuid::common::is_same_model(&microvm_state.vcpu_states[0].cpuid) {
            let state_tsc = microvm_state.vcpu_states[0]
                .tsc_khz
                .ok_or(MicrovmStateError::IncompatibleState(
                    "TSC Error".to_string(),
                ))
                .map_err(RestoreMicrovmState)?;

            if vcpus[0].kvm_vcpu.is_tsc_scaling_required(state_tsc) {
                for vcpu in &vcpus {
                    vcpu.kvm_vcpu
                        .set_tsc_khz(state_tsc)
                        .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
                        .map_err(RestoreMicrovmState)?;
                }
            }
        }
```

Where `is_tsc_scaling_required()` and `set_tsc_khz()` would also consequently change for the better since they would receive the `state_tsc_freq` as a u32 and not an option.

Cosmetically I guess we can go even further. It would be nice if we could do this operation as part of `KvmVcpu::restore_state()`. If it needs to be done earlier, I think we can find a clean way to do it. The implementation details would vary a lot depending on the requirements. We can discuss options offline.

File: src/cpuid/src/common.rs [hunk 15]
```diff
@@ -147,4 +215,46 @@ pub mod tests {
...
     }
+
+    #[test]
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn test_is_same_model() {
+        let kvm = Kvm::new().unwrap();
+        let vm = kvm.create_vm().unwrap();
+        let vcpu = vm.create_vcpu(0).unwrap();
+
+        let mut curr_cpuid = vcpu.get_cpuid2(KVM_MAX_CPUID_ENTRIES).unwrap();
```
Comment (serban300) [inline src/cpuid/src/common.rs:226]: I would create the cpuid manually instead of calling `vcpu.get_cpuid2()`. You can find examples in other unit tests.

**Review by serban300 — CHANGES_REQUESTED** [review]

File: src/cpuid/src/common.rs (hunk 15 above)
Comment (georgepisaltu) [reply src/cpuid/src/common.rs:226]: Done.

File: src/vmm/src/builder.rs (hunk 14 above)
Comment (georgepisaltu) [reply src/vmm/src/builder.rs:450]: While I don't necessarily see it as unsafe, I understand the concern. There is virtually no cost in refactoring this the way you described (in order to get rid of the `unreachable!` statement), so I tried to do it. Let me know if it's better.

Unfortunately, for the last part of the comment, the checks would ideally happen only once, so this means they can't be done inside `KcmVcpu::restore_state()`. This is what we've been trying to avoid. We considered workarounds for this, such as meddling with the TSC `Option<u32>` inside the state, but they feel clunky. @sandreim what do you think?

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 16]
```diff
@@ -15,17 +15,19 @@ use crate::vstate::{
...
     kvm_xsave, CpuId, MsrList, Msrs,
 };
 use kvm_ioctls::{VcpuExit, VcpuFd};
-use logger::{error, IncMetric, METRICS};
+use logger::{error, warn, IncMetric, METRICS};
 use versionize::{VersionMap, Versionize, VersionizeResult};
 use versionize_derive::Versionize;
 use vm_memory::{Address, GuestAddress, GuestMemoryMmap};
 
+const TSC_KHZ_TOL: f64 = 250.0 / 1_000_000.0;
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:29]: `TOL` is the tolerance for normal TSC frequency differences that appear after boot. The calibration process does not yield the same frequency every time. As for the second point, I misplaced the `.round()`. It was supposed to be rounded *after* multiplying with the current value. Fixed it now.

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 17]
```diff
@@ -291,9 +305,69 @@ impl KvmVcpu {
...
+                    Ok(false)
+                }
+            }
+            None => {
+                // We don't know the TSC freq for this machine,
+                // but we still have the one in the state. Try
+                // to set that regardless as a last effort.
+                Ok(true)
+            }
+        }
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs:352]: After the refactoring for the other comment, this is not needed anymore.

File: src/vmm/src/builder.rs [hunk 18]
```diff
@@ -427,6 +429,36 @@ pub fn build_microvm_from_snapshot(
...
+        // Scale the TSC frequency for all VCPUs, if needed.
+        if vcpus[0].kvm_vcpu.is_tsc_scaling_required(state_tsc) {
+            for vcpu in &vcpus {
+                vcpu.kvm_vcpu
+                    .set_tsc_khz(state_tsc)
+                    .map_err(|e| MicrovmStateError::IncompatibleState(e.to_string()))
+                    .map_err(RestoreMicrovmState)?;
+            }
+        }
+    }
```
Comment (serban300) [inline src/vmm/src/builder.rs:460]: I would move this code to the beginning of `restore_vcpu_states()`

File: tests/integration_tests/functional/test_snapshot_advanced.py [hunk 19]
```diff
@@ -137,6 +137,58 @@ def test_restore_old_version_all_devices(bin_cloner_path):
...
+        fc_binary=firecracker.local_path(),
+        jailer_binary=jailer.local_path(),
+        diff_snapshots=diff_snapshots,
+        balloon=True
+    )
+
+    # Resume microvm using current build of FC/Jailer.
+    # The resume should be successful because the CPU model
+    # in the snapshot state is the same as this host's.
+    microvm, _ = builder.build_from_snapshot(
```
Comment (sandreim) [inline tests/integration_tests/functional/test_snapshot_advanced.py:176]: Please also add an additional issue to cover negative testing on different CPU models.

File: src/vmm/src/builder.rs (hunk 14 above)
Comment (sandreim) [reply src/vmm/src/builder.rs]: Failing early and reducing resume cost are good reasons to have only one of the CPU states checked once. This works with the simplifying assumption that the host supports invariant TSC and freq is the same across all CPUs.

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 20]
```diff
@@ -392,6 +434,37 @@ pub struct VcpuState {
...
+}
+
+impl VcpuState {
+    fn default_tsc_khz(_: u16) -> Option<u32> {
+        warn!("CPU TSC freq not found in snapshot");
+        None
+    }
+
+    fn ser_tsc(&mut self, target_version: u16) -> VersionizeResult<()> {
+        if target_version < 2 {
```
Comment (sandreim) [inline src/vmm/src/vstate/vcpu/x86_64.rs:448]: This can be simplified, no need to check the target_version here, please see https://docs.rs/versionize_derive/0.1.4/versionize_derive/derive.Versionize.html#ser_fn

Also I would add asserts in tests for the 2 warnings.

File: src/vmm/src/builder.rs (hunk 18 above)
Comment (serban300) [reply src/vmm/src/builder.rs:460]: When Firecracker is used correctly, we shouldn't encounter issues related to setting the TSC. Failing should be more of an edge case. So I think that it's worth the move for readability.

Comment (georgepisaltu) [reply src/vmm/src/builder.rs:460]: The problem is that by the time `restore_vcpu_states` is called, the VCPUs are already moved to their own threads and we run into the same problem as before regarding the redundant checks. Furthermore, it would just add extra complexity to the code to try and send these check specific fields through the `VcpuHandle` interface. I lean towards keeping it as is.

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 21]
```diff
@@ -291,9 +306,35 @@ impl KvmVcpu {
...
+            // per million beacuse it is common for TSC frequency
+            // to differ due to calibration at boot time.
+            Some(curr_tsc_freq) => {
+                let diff = (curr_tsc_freq as i64 - state_tsc_freq as i64).abs();
+                diff > (state_tsc_freq as f64 * TSC_KHZ_TOL).round() as i64
+            }
+            // We don't know the TSC freq for this machine,
+            // but we still have the one in the state. Try
+            // to set that regardless as a last effort.
+            None => true,
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs]: Refactored to fail when `get_tsc_khz` fails.

File: src/vmm/src/vstate/vcpu/x86_64.rs [hunk 22]
```diff
@@ -428,6 +503,31 @@ mod tests {
         (vm, vcpu, vm_mem)
     }
 
+    /*fn mutate_state_cpuid_vendor(orig_state: &VcpuState) -> VcpuState {
```
[the remark this answers is not in the document]
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs]: Done.

File: tests/integration_tests/functional/test_snapshot_advanced.py (hunk 19 above)
Comment (georgepisaltu) [reply tests/integration_tests/functional/test_snapshot_advanced.py:176]: Will do, after this PR is merged.

File: src/vmm/src/vstate/vcpu/x86_64.rs (hunk 20 above)
Comment (georgepisaltu) [reply src/vmm/src/vstate/vcpu/x86_64.rs]: I refactored this and added a negative test for the second message. The first one is really tough to get. I left a TODO there for when we decide to have mock structures for testing here, so we can easily cover the case when `get_tsc_khz` doesn't work.

**Review by georgepisaltu — COMMENTED** [review]
@serban300 @sandreim PTAL!

**Review by serban300 — APPROVED** [review]

**Review by sandreim — APPROVED** [review]
